### PR TITLE
Extend Transformers Trainer Class to Enable PyTorch Torchscript for Inference

### PR DESCRIPTION
**Pull request by jianan-gu**
# What does this PR do?

This PR intends to extend Transformers Trainer Class with PyTorch [Torchscript](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) (`torch.jit.trace`) to speed up model inference with just-in-time compilation.

Users could simply enable "--jit_mode" Trainer input args to get benifits.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by lewtun** [comment]
> This may be more [optimum](https://github.com/huggingface/optimum/)'s domain than Transformers, what do you think @mfuntowicz @LysandreJik @lewtun ?

Thanks for the ping and thank you @jianan-gu for opening the PR! 

I agree that inference-based improvements are better suited in `optimum`, so in this case we'd need to implement a dedicated `torchscript` subpackage that effectively extends the `Trainer` in this manner. 

Having said that, I wonder if it's overkill to create a new package just to support this feature, which is natively included in PyTorch. 

@mfuntowicz @michaelbenayoun do you see any other benefits that could be had by having a dedicated subpackage for `torchscript` in `optimum`?

**Comment by stas00** [comment]
I'd be happy to work on this PR if you give me a green light that you'll accept this feature.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
 
         return model
 
-    def _wrap_model(self, model, training=True):
+    def torch_jit_model(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = ()
+            for _, batch in enumerate(dataloader):
+                for _, label in enumerate(batch):
```
Comment (sgugger) [inline src/transformers/trainer.py:1103]: I'm not sure I understand what is happening here. The label name is not necessarily the last key in the batch, there is a `self.label_names` attribute for this.

Also why use `enumerate` if it's to discard the index?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
...
+                    dumpy_tensor = torch.ones_like(batch[label])
+                    L1 = list(jit_inputs)
+                    L1.append(dumpy_tensor)
```
Comment (sgugger) [inline src/transformers/trainer.py:1106]: Please use a better variable name that is more descriptive of what `L1` is.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
...
+                break
+            try:
+                if self.args.bf16:
+                    with self.autocast_smart_context_manager(), torch.no_grad():
+                        model = torch.jit.trace(model, jit_inputs, strict=False)
+                    model = torch.jit.freeze(model)
+                else:
+                    with torch.no_grad():
+                        model = torch.jit.trace(model, jit_inputs, strict=False)
+                    model = torch.jit.freeze(model)
```
Comment (sgugger) [inline src/transformers/trainer.py:1117]: Let's group all of this in one line without test. We can build the proper list of context managers (`[torch.no_grad()]` and then append to it the other one if needed) and wrap it in `ContextManagers` (that you can find in utils).

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -234,6 +234,8 @@ class TrainingArguments:
             Random seed to be used with data samplers. If not set, random generators for data sampling will use the
             same seed as `seed`. This can be used to ensure reproducibility of data sampling, independent of the model
             seed.
+        jit_mode(`bool`, *optional*, defaults to `False`):
+            Try to use PyTorch jit trace for inference.
```
Comment (sgugger) [inline src/transformers/training_args.py:238]:
```suggestion
            Whether ot not to use PyTorch jit trace for inference.
```

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -583,6 +585,7 @@ class TrainingArguments:
     no_cuda: bool = field(default=False, metadata={"help": "Do not use CUDA even when it is available"})
     seed: int = field(default=42, metadata={"help": "Random seed that will be set at the beginning of training."})
     data_seed: Optional[int] = field(default=None, metadata={"help": "Random seed to be used with data samplers."})
+    jit_mode: bool = field(default=False, metadata={"help": "Try to use PyTorch jit trace for inference"})
```
Comment (sgugger) [inline src/transformers/training_args.py:588]:
```suggestion
    jit_mode: bool = field(default=False, metadata={"help": "Whether or not to use PyTorch jit trace for inference"})
```

**Review by sgugger — COMMENTED** [review]
Thanks for adding this. We would need some test before we can merge the PR, and the code needs to be adapted to the way the labels are handled in Transformers.

File: src/transformers/trainer.py (hunk 2 above)
Comment (jianan-gu) [reply src/transformers/trainer.py]: `L1 `was a temp list, here made changes to just use `jit_inputs` list  to avoid confusing, thanks.

File: src/transformers/trainer.py (hunk 3 above)
Comment (jianan-gu) [reply src/transformers/trainer.py]: Have made changes to refine the code accordingly, thanks.

File: src/transformers/trainer.py (hunk 1 above)
Comment (jianan-gu) [reply src/transformers/trainer.py]: Here what we do is to construct the `jit_inputs `for `torch.jit.trace`, as steps:
(1) get the first batch from the dataloader as the `example_batch`
(2) enumerate each tensor from this `example_batch` and create each `example_tensor`
(3) tuple the `jit_inputs ` and it will be used in `torch.jit.trace`.

Therefore we do not need `self.label_names`, we are enumerating the datasets tensors ( batch: {key : tensor}, not related to labels)

Thanks.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1140,7 +1141,28 @@ def call_model_init(self, trial=None):
 
         return model
 
-    def _wrap_model(self, model, training=True):
+    def torch_jit_model(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            example_batch = list(dataloader)[0]
```
Comment (sgugger) [inline src/transformers/trainer.py:1147]: No, we can't unpack the whole dataloader as it might be very slow and make the user run out of RAM if the dataset is huge.
```suggestion
            for example_batch in dataloader:
                break
```

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -1140,7 +1141,28 @@ def call_model_init(self, trial=None):
...
+            for _, key in enumerate(example_batch):
```
Comment (sgugger) [inline src/transformers/trainer.py:1148]:
```suggestion
            for key in example_batch:
```
Again, why enumerate when the index is not used?

**Review by sgugger — COMMENTED** [review]
Thanks for addressing the comments, this will also need a test before we can merge.

**Comment by jianan-gu** [comment]
Adding UT Tests in test_trainer.py for jit which covers `evaluate `and `predict`.
Thanks.

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
...
+            jit_inputs = tuple(jit_inputs)
+            try:
+                jit_model = model.eval()
+                with ContextManagers([self.autocast_smart_context_manager(), torch.no_grad()]):
+                    jit_model = torch.jit.trace(jit_model, jit_inputs, strict=False)
+                jit_model = torch.jit.freeze(jit_model)
+                jit_model(**example_batch)
+                model = jit_model
+            except (RuntimeError, TypeError):
+                logger.warning("fail to use PyTorch jit mode")
```
Comment (sgugger) [inline src/transformers/trainer.py:1178]: Maybe include the error message in the warning here?

File: src/transformers/training_args.py [hunk 9]
```diff
@@ -236,6 +236,8 @@ class TrainingArguments:
             Random seed to be used with data samplers. If not set, random generators for data sampling will use the
             same seed as `seed`. This can be used to ensure reproducibility of data sampling, independent of the model
             seed.
+        jit_mode(`bool`, *optional*, defaults to `False`):
```
Comment (sgugger) [inline src/transformers/training_args.py:239]: We should maybe add an `eval` suffix here to make it clear to the user this is inference only.

**Review by sgugger — APPROVED** [review]
Thanks for adding the tests! I have a few last small comments.

Do you want to have a second look @stas00 ?

File: src/transformers/training_args.py (hunk 9 above)
Comment (jianan-gu) [reply src/transformers/training_args.py]: sure, have made changes, thanks for the suggestion.

File: src/transformers/trainer.py (hunk 8 above)
Comment (jianan-gu) [reply src/transformers/trainer.py]: have made changes accordingly, thanks.

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
 
         return model
 
+    def torch_jit_model(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            for example_batch in dataloader:
+                break
```
Comment (stas00) [inline src/transformers/trainer.py:1165]:
```suggestion
            example_batch = next(iter(dataloader))
```

perhaps, otherwise it reads a bit as a beginner python code.

but also `dataloader` can be `None` , so this code needs to be more defensive.

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
...
+            jit_inputs = tuple(jit_inputs)
+            try:
+                jit_model = model.eval()
+                with ContextManagers([self.autocast_smart_context_manager(), torch.no_grad()]):
+                    jit_model = torch.jit.trace(jit_model, jit_inputs, strict=False)
+                jit_model = torch.jit.freeze(jit_model)
+                jit_model(**example_batch)
+                model = jit_model
+            except (RuntimeError, TypeError) as ex:
+                logger.warning(f"fail to use PyTorch jit mode due to: {ex}.")
```
Comment (stas00) [inline src/transformers/trainer.py:1178]:
```suggestion
            except (RuntimeError, TypeError) as exception:
                logger.warning(f"failed to use PyTorch jit mode due to: {exception}.")
```

**Review by stas00 — CHANGES_REQUESTED** [review]
What is the sounds of a tree falling in the forest when there is nobody to hear it?

All these new features need to be documented on the user-side so that they actually get used. So the same as with IPEX let's add user docs and as before ideally a small benchmark to give users an incentive to try the feature.

updade: I added a doc for inference as we haven't created it and then you can push the usage examples there - like we did with IPEX PR. So please fill out the blank and then we are good to go I think.

and added a few nits in the code.

File: docs/source/en/perf_infer_gpu_one.mdx [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+-->
+
+# Efficient Inference on a Single GPU
+
+This guide focuses on inferencing large models efficiently on a single GPU.
+
+## JIT-mode
+
+XXX: fill in the details
```
[the remark this answers is not in the document]
Comment (jianan-gu) [reply docs/source/en/perf_infer_gpu_one.mdx:18]: Yes, it is for inference only.

[the remark this answers is not in the document]
Comment (stas00) [reply docs/source/en/perf_infer_gpu_one.mdx:18]: so why `torch_jit_model` name? let's name it accordingly?

[the remark this answers is not in the document]
Comment (jianan-gu) [reply docs/source/en/perf_infer_gpu_one.mdx:18]: Sure, have made changes accordingly, thanks for your suggestion.

[the remark this answers is not in the document]
Comment (stas00) [reply docs/source/en/perf_infer_gpu_one.mdx:18]: Thank you, @jianan-gu - please ping me when it's ready for a final review.

**Comment by jianan-gu** [comment]
> What is the sounds of a tree falling in the forest when there is nobody to hear it?
> 
> All these new features need to be documented on the user-side so that they actually get used. So the same as with IPEX let's add user docs and as before ideally a small benchmark to give users an incentive to try the feature.
> 
> updade: I added a doc for inference as we haven't created it and then you can push the usage examples there - like we did with IPEX PR. So please fill out the blank and then we are good to go I think.
> 
> and added a few nits in the code.

Though this jit mode works both for CPU and GPU, the current IPEX release covers CPU side optimizations with jit mode for model inference. Therefore we only have added and updated the inference doc for CPU (perf_infer_cpu.mdx), please take a review of the contents @stas00 , thanks. (for the small benchmark, for now, we only have relative performance numbers like shown in https://github.com/huggingface/transformers/issues/17137, so we will prepare for the numbers with --skip_memory_metrics 0 as a follow-up)

File: docs/source/en/perf_infer_cpu.mdx [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+For a gentle introduction to TorchScript, see the Introduction to [PyTorch TorchScript tutorial](https://pytorch.org/tutorials/beginner/Intro_to_TorchScript_tutorial.html#tracing-modules).
+
+### IPEX Graph Optimization with JIT-mode
+Intel® Extension for PyTorch provides further optimizations in jit mode for Transformers series models. It is highly recommended for users to take advantage of Intel® Extension for PyTorch with jit mode. Some frequently used operator patterns from Transformers models are already supported in Intel® Extension for PyTorch with jit mode fusions. Those fusion patterns like Multi-head-attention fusion, Concat Linear, Linear+Add, Linear+Gelu, Add+LayerNorm fusion and etc, are enabled and perform well. The benefit of the fusions are delivered to users in a transparant fashion. According to the analysis, ~70% of most popular NLP tasks in question-answering, text-classification, and token-classification can get performance benefits with these fusion patterns for both Float32 precision and BFloat16 Mixed precision.
```
Comment (stas00) [inline docs/source/en/perf_infer_cpu.mdx:23]:
```suggestion
Intel® Extension for PyTorch provides further optimizations in jit mode for Transformers series models. It is highly recommended for users to take advantage of Intel® Extension for PyTorch with jit mode. Some frequently used operator patterns from Transformers models are already supported in Intel® Extension for PyTorch with jit mode fusions. Those fusion patterns like Multi-head-attention fusion, Concat Linear, Linear+Add, Linear+Gelu, Add+LayerNorm fusion and etc. are enabled and perform well. The benefit of the fusion is delivered to users in a transparent fashion. According to the analysis, ~70% of most popular NLP tasks in question-answering, text-classification, and token-classification can get performance benefits with these fusion patterns for both Float32 precision and BFloat16 Mixed precision.
```

File: src/transformers/trainer.py [hunk 14]
```diff
@@ -1158,6 +1158,26 @@ def call_model_init(self, trial=None):
 
         return model
 
+    def torch_jit_model_eval(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            example_batch = next(iter(dataloader))
```
Comment (stas00) [inline src/transformers/trainer.py:1164]: is `dataloader` required? As I commented before you're not checking anywhere if it's None. 

If it's required then please remove `=None` in the function signature.

**Review by stas00 — CHANGES_REQUESTED** [review]
last few small things to adjust and then it should be good.

File: src/transformers/trainer.py (hunk 14 above)
Comment (jianan-gu) [reply src/transformers/trainer.py]: yes, it is required for jit.
Thanks for your suggestion, here have made changes to add a checking before constructing the jit inputs and removed `=None` in the function signature.

Comment (stas00) [reply src/transformers/trainer.py]: All looks good now to me.

Thank you!

**Comment by stas00** [comment]
@sgugger, would you like to have a quick look at the last changes since your review - mainly the newly added doc. Thank you!

otherwise it's good to be merged.

**Review by stas00 — APPROVED** [review]

File: docs/source/en/perf_infer_cpu.mdx [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+- Inference using jit mode on CPU:
+<pre> python run_qa.py \
+--model_name_or_path csarron/bert-base-uncased-squad-v1 \
+--dataset_name squad \
+--do_eval \
+--max_seq_length 384 \
+--doc_stride 128 \
+--output_dir /tmp/ \
+--no_cuda \
+<b> --jit_mode_eval </b></pre>
```
Comment (sgugger) [inline docs/source/en/perf_infer_cpu.mdx:45]: Not a big fan of using the HTML syntax here, I'd prefer the triple backticks, and it's okay if the new arg is not in bold. If keeping it, some spaces need to be removed.
```suggestion
<pre>python run_qa.py \
--model_name_or_path csarron/bert-base-uncased-squad-v1 \
--dataset_name squad \
--do_eval \
--max_seq_length 384 \
--doc_stride 128 \
--output_dir /tmp/ \
--no_cuda \
<b>--jit_mode_eval </b></pre> 
```

File: docs/source/en/perf_infer_cpu.mdx [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+<pre> python run_qa.py \
+--model_name_or_path csarron/bert-base-uncased-squad-v1 \
+--dataset_name squad \
+--do_eval \
+--max_seq_length 384 \
+--doc_stride 128 \
+--output_dir /tmp/ \
+--no_cuda \
+<b> --use_ipex \ </b>
+<b> --jit_mode_eval </b></pre>
```
Comment (sgugger) [inline docs/source/en/perf_infer_cpu.mdx:57]: Same comment
```suggestion
<pre>python run_qa.py \
--model_name_or_path csarron/bert-base-uncased-squad-v1 \
--dataset_name squad \
--do_eval \
--max_seq_length 384 \
--doc_stride 128 \
--output_dir /tmp/ \
--no_cuda \
<b>--use_ipex \</b>
<b>--jit_mode_eval</b></pre> 
```

**Review by sgugger — APPROVED** [review]
Some small comments on the new doc, thanks for adding it!
